src/http_rest_api_handler.cpp

```diff
@@ -204,24 +204,227 @@ std::string HttpRestApiHandler::preprocessInferRequest(std::string request_body)
+    for (int i = 0; i < grpc_request.mutable_inputs()->size(); i++) {
+        auto input = grpc_request.mutable_inputs()->Mutable(i);
+        auto status = validateContentFieldsEmptiness(input);
+        if (!status.ok()) {
+            return status;
+        }
+        if (input->parameters().count("binary_data_size") > 0) {
+            if (input->parameters().at("binary_data_size").parameter_choice_case() == inference::InferParameter::ParameterChoiceCase::kInt64Param) {
+                auto binary_input_size = input->parameters().at("binary_data_size").int64_param();
+                if (binary_input_offset + binary_input_size > binary_inputs_size) {
+                    SPDLOG_ERROR("Binary inputs size exceeds provided buffer size {}", binary_inputs_size);
+                    return StatusCode::REST_BINARY_BUFFER_ECEEDED;
+                }
+                status = parseBinaryInput(input, binary_input_size, binary_inputs + binary_input_offset);
+                if (!status.ok()) {
+                    return status;
+                }
+                binary_input_offset += binary_input_size;
+            } else if (input->parameters().at("binary_data_size").parameter_choice_case() == inference::InferParameter::ParameterChoiceCase::kStringParam) {
+                std::vector<int> binary_inputs_sizes;
+                status = convertStringToVectorOfSizes(input->parameters().at("binary_data_size").string_param(), binary_inputs_sizes);
+                if (!status.ok()) {
+                    return status;
+                }
+                for (auto size : binary_inputs_sizes) {
+                    if (binary_input_offset + size > binary_inputs_size) {
+                        SPDLOG_ERROR("Binary inputs size exceeds provided buffor size {}", binary_inputs_size);
+                        return StatusCode::REST_BINARY_BUFFER_ECEEDED;
+                    }
+                    status = parseBinaryInput(input, size, binary_inputs + binary_input_offset);
+                    if (!status.ok()) {
+                        return status;
+                    }
+                    binary_input_offset += size;
+                }
+            } else {
+                SPDLOG_WARN("binary_data_size parameter type should be int64 or string");
+                return StatusCode::REST_BINARY_DATA_SIZE_PARAMETER_INVALID;
+            }
+        }
+    }
+    return StatusCode::OK;
+}
+
+Status HttpRestApiHandler::prepareGrpcRequest(const std::string modelName, const std::string modelVersion, const std::string request_body, ::inference::ModelInferRequest& grpc_request) {
     google::protobuf::util::JsonParseOptions opts;
     opts.ignore_unknown_fields = true;
-    google::protobuf::util::JsonStringToMessage(request, &grpc_request, opts);
+    auto jsonClosingBracket = request_body.find_last_of("}");
+    bool dataAfterHeader = ((jsonClosingBracket != std::string::npos) && (jsonClosingBracket != (request_body.length() - 1)));
```

Comment:
Here we should add unit to ensure we never buffer overflow. Can we somehow mock such situation?